site/content/blog/2020-05-20-hands-on-iouring-go.md

```diff
@@ -0,0 +1,327 @@
+
+Fortunately, Jens Axboe, the creator of `io_uring` has created a wrapper library called `liburing` to help simplify all of this. With `liburing`, we roughly have to do this set of steps:
+
+- `io_uring_queue_(init|exit)` to setup and tear down the io_uring.
+
+- `io_uring_get_sqe` to get an SQE.
+
+- `io_uring_prep_(readv|writev|other)` to mark which syscall to use.
+
+- `io_uring_sqe_set_data` to mark the user data field.
+
+- `io_uring_(wait|peek)_cqe` to either wait for cqe or peek for it without waiting.
+
+- `io_uring_cqe_get_data` to get back the user data field.
+
+- `io_uring_cqe_seen` to mark the CQE as done.
+
+### Wrapping io_uring in Go
+
+This was a lot of theory to digest. There's even more of it which I have deliberately skipped for the sake of brevity. Now let's get back to writing some code in Go and take this for a spin. For simplicity and safety we will use the `liburing` library, which means we will need to use CGo. That's fine because this is just a toy, and the right way would be have [native support](https://github.com/golang/go/issues/31908) in the Go runtime. As a result of that, we will unfortunately have to use callbacks. In native Go, the running goroutine would be put to sleep by the runtime, and then woken up when the data would be available in the completion queue.
+
+Let's name our package `frodo` (and just like that I have knocked out one of the two hardest problems in computer science). We will just have a very simple API to read and write files. And two more functions to setup and cleanup the ring when done.
+
+Our main workhorse will be a single goroutine which takes in submission requests and pushes them to the SQ. And from C we will make a callback to Go with the CQE entry. We will use the `fd` of the files to know which callback to execute once we get our data. However, we also need to decide when to actually submit the queue to the kernel. We maintain a queue threshold, where if we exceed a threshold  of pending requests, we submit. And also, we expose another function to the user to allow them to make the submission by themselves to allow them to have better control over application behavior.
+
+Note that, this again is an inefficient way of doing things. Since the CQ and SQ are completely separate they don't need any sort of locking at all, and therefore submission and completion can happen freely from different threads. Ideally, we would just push an entry to SQ, and have a separate goroutine listening for completion waiting, and whenever we see an entry we make a callback and go back to waiting. Remember that we can use `io_uring_enter` to do just completion ? Here is one such example ! This still makes one syscall per CQE entry, and we can even optimize it further by specifying number of CQE entries to wait for.
```

Comment:
```suggestion
Note that, this again is an inefficient way of doing things. Since the CQ and SQ are completely separate they don't need any sort of locking at all, and therefore submission and completion can happen freely from different threads. Ideally, we would just push an entry to SQ, and have a separate goroutine listening for completion waiting, and whenever we see an entry we make a callback and go back to waiting. Remember that we can use `io_uring_enter` to do just completion? Here is one such example! This still makes one syscall per CQE entry, and we can even optimize it further by specifying number of CQE entries to wait for.
```